### PR TITLE
[Merged by Bors] - Implement `WeakMap`

### PR DESCRIPTION
**Pull request by raskad**
This Pull Request changes the following:

- Implement `WeakMap` buildin object.

### CODE REVIEW

File: boa_engine/src/builtins/weak_map/mod.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        // 3. Let entries be M.[[WeakMapData]].
+        // 4. If key is not an Object, return undefined.
+        let Some(key) = args.get_or_undefined(0).as_object() else {
+            return Ok(JsValue::undefined());
+        };
+
+        // 5. For each Record { [[Key]], [[Value]] } p of entries, do
+        // a. If p.[[Key]] is not empty and SameValue(p.[[Key]], key) is true, return p.[[Value]].
+        // 6. Return undefined.
+        if let Some(value) = m.get(key.inner()) {
```
Comment (jedel1043) [inline boa_engine/src/builtins/weak_map/mod.rs:184]: Could use `unwrap_or_default` instead.

**Review by jedel1043 — APPROVED** [review]
Nice job! Just had a small nitpick which doesn't block merging this.

**Comment by jasonwilliams** [comment]
I do want to take a look at this when I get some time, looks good so far though

**Review by jasonwilliams — APPROVED** [review]
LGTM

**Pull request closed without merging**